manubot/cite/handlers.py

```diff
@@ -0,0 +1,811 @@
+import abc
+import dataclasses
+import functools
+import re
+import typing
+
+from manubot.util import import_function
+
+"""
+Non-citation prefixes used by the pandoc-xnos suite of Pandoc filters,
+including pandoc-fignos, pandoc-tablenos, and pandoc-eqnos.
+"""
+_pandoc_xnos_prefixes = {"fig", "tbl", "eq"}
+
+
+_local_handlers = [
+    "manubot.cite.arxiv.Handler_arXiv",
+    "manubot.cite.doi.Handler_DOI",
+    "manubot.cite.isbn.Handler_ISBN",
+    "manubot.cite.pubmed.Handler_PMC",
+    "manubot.cite.pubmed.Handler_PubMed",
+    "manubot.cite.url.Handler_URL",
+    "manubot.cite.wikidata.Handler_Wikidata",
+]
+
+
+@functools.lru_cache(maxsize=10_000)
+def get_handler(prefix_lower):
+    if not isinstance(prefix_lower, str):
+        raise TypeError(
+            f"prefix_lower should be a str, instead received {prefix_lower.__class__.__name__}"
+        )
+    assert prefix_lower == prefix_lower.lower()
+    handler = prefix_to_handler[prefix_lower]
+    handler = import_function(handler)(prefix_lower)
+    return handler
+
+
+def _generate_prefix_to_handler() -> typing.Dict[str, str]:
+    """
+    Generate complete dictionary for prefix_to_handler.
+    """
+    import inspect
+
+    from .curie import get_curie_handlers
+
+    curie_handlers = get_curie_handlers()
+    pth = {}
+    for handler in curie_handlers + _local_handlers:
+        if isinstance(handler, str):
+            handler = import_function(handler)("dummy_prefix_lower")
+        for prefix in handler.prefixes:
+            pth[
+                prefix
+            ] = f"{inspect.getmodule(handler).__name__}.{handler.__class__.__name__}"
+    pth = dict(sorted(pth.items()))  # sort
```

Comment:
No. sorting is to keep the diff minimal when more handlers are added. Since pth is a dict, if prefix is used multiple times, whichever one is defined last will remain.